File: sender.py
import socket
from Cryptodome.Cipher import AES
from Cryptodome.Util.Padding import pad, unpad
import base64

# 输入接收端域名
receiver_host = input("请输入接收端域名：")

# 输入密钥
key = input("请输入密钥：").encode()
key = key.ljust(16, b'0')[:16]  # 用0填充至16位

# 创建AES加密器
cipher = AES.new(key, AES.MODE_ECB)

# 第一次握手
receiver_port = 12345
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.connect((receiver_host, receiver_port))
sock.sendall(b'Hello')
response = sock.recv(1024)
if response == b'ACK':
    print("第一次握手成功！")
else:
    print("第一次握手失败！")
    sock.close()
    exit(1)

while True:
    # 加密并发送消息
    message = input("请输入要发送的消息（输入'exit'退出）：")
    if message == 'exit':
        break

    padded_message = pad(message.encode(), AES.block_size)
    encrypted_message = cipher.encrypt(padded_message)
    encoded_message = base64.b64encode(encrypted_message)
    sock.sendall(encoded_message)

    # 接收并解密回复消息
    response = sock.recv(1024)
    encoded_response = base64.b64decode(response)
    decrypted_response = cipher.decrypt(encoded_response)
    unpadded_response = unpad(decrypted_response, AES.block_size)
    print("接收到的回复消息：", unpadded_response.decode())

# 关闭连接
sock.close()

File: Sender.java
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.util.Arrays;

public class Sender {
    private static final int RECEIVER_PORT = 12345;
    private static final int BUFFER_SIZE = 1024;

    public static void main(String[] args) {
        try {
            // 输入接收端域名
            System.out.print("请输入接收端域名：");
            String receiverHost = System.console().readLine();

            // 输入密钥
            System.out.print("请输入密钥：（输入时自动隐去输入内容）");
            char[] keyChars = System.console().readPassword();
            byte[] key = new String(keyChars).getBytes(StandardCharsets.UTF_8);
            key = padKey(key);

            // 创建AES加密器
            Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding");
            SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);

            // 第一次握手
            Socket socket = new Socket(receiverHost, RECEIVER_PORT);
            OutputStream outputStream = socket.getOutputStream();
            InputStream inputStream = socket.getInputStream();

            outputStream.write("Hello".getBytes(StandardCharsets.UTF_8));
            byte[] buffer = new byte[BUFFER_SIZE];
            int bytesRead = inputStream.read(buffer);
            String response = new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);
            if (response.equals("ACK")) {
                System.out.println("第一次握手成功！");
            } else {
                System.out.println("第一次握手失败！");
                socket.close();
                System.exit(1);
            }

            while (true) {
                // 加密并发送消息
                System.out.print("请输入要发送的消息（输入'exit'退出）：");
                String message = System.console().readLine();
                if (message.equals("exit")) {
                    break;
                }

                byte[] paddedMessage = padMessage(message.getBytes(StandardCharsets.UTF_8));
                byte[] encryptedMessage = cipher.doFinal(paddedMessage);
                byte[] encodedMessage = Base64.getEncoder().encode(encryptedMessage);
                outputStream.write(encodedMessage);

                // 接收并解密回复消息
                bytesRead = inputStream.read(buffer);
                byte[] partialBuffer = Arrays.copyOfRange(buffer, 0, bytesRead);
                byte[] encodedResponse = Base64.getDecoder().decode(partialBuffer);
                byte[] decryptedResponse = cipher.doFinal(encodedResponse);
                String unpaddedResponse = new String(decryptedResponse, StandardCharsets.UTF_8).trim();
                System.out.println("接收到的回复消息：" + unpaddedResponse);
            }

            // 关闭连接
            socket.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static byte[] padKey(byte[] key) {
        byte[] paddedKey = new byte[16];
        System.arraycopy(key, 0, paddedKey, 0, Math.min(key.length, paddedKey.length));
        return paddedKey;
    }

    private static byte[] padMessage(byte[] message) {
        int paddingLength = 16 - (message.length % 16);
        byte[] paddedMessage = new byte[message.length + paddingLength];
        System.arraycopy(message, 0, paddedMessage, 0, message.length);
        return paddedMessage;
    }
}
